File: 108-convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* subTree(TreeNode* root, vector<int> &nums, int l, int r){
        if(l > r) return root;
        int mid = (l+r)/2;
        TreeNode* temp = new TreeNode(nums[mid]);
        root = temp;
        
        temp->left = subTree(root->left, nums, l, mid-1);
        temp->right = subTree(root->right, nums, mid+1, r);

        return root;

    }
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        TreeNode* root = nullptr;
        root = subTree(root, nums, 0, nums.size()-1);

        return root;
    }
};

File: 108-convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def subTree(nums: List[int], l: int, r: int) -> Optional[TreeNode]:
            if l>r:
                return None
            mid = (l+r)//2
            node = TreeNode(nums[mid])
            node.left = subTree(nums, l, mid-1)
            node.right = subTree(nums, mid+1, r)

            return node
        
        return subTree(nums, 0, len(nums)-1)